common/robotPositioner.h

```diff
@@ -0,0 +1,186 @@
+    millimeter_t m_robot_D;                                         // robots' axis length
+    millimeter_t m_stopping_distance;                               // if the robot's distance from goal < stopping dist, robot stops
+    degree_t m_allowed_heading_error;                               // the amount of error allowed in the final heading
+    double m_k1;                                                    // curveness of the path to the goal
+    double m_k2;                                                    // speed of turning on the curves
+    double m_alpha;                                                 // causes more sharply peaked curves 
+    double m_beta;                                                  // causes to drop velocity if 'k'(curveness) increases
+    meters_per_second_t m_max_velocity;                             // max velocity
+
+    // robot superclass
+    BoBRobotics::Robots::Tank &m_bot;   
+  
+
+    
+    void driveRobotWithVelocities(const meters_per_second_t v,
+                                  const degrees_per_second_t w,
+                                  const millimeter_t robot_r,
+                                  const millimeter_t robot_d) {
+        // v = r * (Vl+Vr)/2
+        // w = r * (Vr-Vl)/D
+        double a =  (robot_r/2).value(); //convert to cm
+        double b =  (robot_r/2).value();
+        double c = ( robot_r / robot_d).value();
+        double d = (-(robot_r / robot_d).value());
+
+        double e=(v.value()); // converting to millimeters/second 
+        double f=w.value();
+        double Vl,Vr;
+
+        // determinant
+        double det = (a*d - b*c);
+    
+        Vl = ((e*d - b*f)/det);
+        Vr = ((a*f - e*c)/det);
+        
+        // drive robot 
+        m_bot.tank(Vl, Vr);
+    }
+
+    
+    degree_t angleWrapAround(degree_t angle) {
+        // changing from <0,360> to <-180, 180>
+        if (angle <= degree_t(0)) { angle += degree_t(360); } 
+        if (angle > degree_t(180)) { angle -= degree_t(360); }
+        return angle;
+    }
+
+
+    template <class T, class S>
+    static bool closeTo(const T num, const S threshold) {
+        return (num > -threshold && num < threshold);
+    }
+
+    // updates the range and bearing from the goal location
+    void updateRangeAndBearing() {
+        
+        millimeter_t delta_x = m_goalPositionX - m_pos_X;
+        millimeter_t delta_y = m_goalPositionY - m_pos_Y;
+    
+        // calculate distance
+        m_distanceFromGoal =  units::math::hypot(delta_x, delta_y);
+
+        // calculate bearing
+        m_bearingFromGoal = units::math::atan2(delta_y,delta_x)-m_heading;
+       
+        // changing from <0,360> to <-180, 180>
+        m_bearingFromGoal = angleWrapAround(m_bearingFromGoal);
+        
+    } 
+
+//-----------------PUBLIC API---------------------------------------------------------------------
+public:
+
+    RobotPositioner(
+        BoBRobotics::Robots::Tank &tank,
```

Comment:
This is a much nicer interface - my only thought is, how about you add virtuals for ``millimeter_t getWheelRadius() const`` and ``millimeter_t getAxisLength() const`` to ``BoBRobotics::Robots::Tank`` - I have a Surveyor on my desk if you wanna make whatever measurements are required

Reply:
That's an excellent idea!